astropy/units/format/base.py

```diff
@@ -151,34 +149,46 @@ def to_string(
         # First the scale.  Normally unity, in which case we omit
         # it, but non-unity scale can happen, e.g., in decompositions
         # like u.Ry.decompose(), which gives "2.17987e-18 kg m2 / s2".
-        if unit.scale == 1:
-            s = ""
-        else:
-            s = cls.format_exponential_notation(unit.scale)
-
-        # Now the unit baes, taking care that dimensionless does not have any
-        # (but can have a scale; e.g., u.percent.decompose() gives "0.01").
-        if len(unit.bases):
-            if s:
-                s += cls._scale_unit_separator
-            if fraction:
-                numerator, denominator = utils.get_grouped_by_powers(
-                    unit.bases, unit.powers
-                )
-            else:
-                numerator = list(zip(unit.bases, unit.powers))
-                denominator = []
-            if len(denominator):
-                if len(numerator):
-                    numerator = cls._format_unit_list(numerator)
-                else:
-                    numerator = "1"
-                denominator = cls._format_unit_list(denominator)
-                s = cls._format_fraction(s, numerator, denominator, fraction=fraction)
+        s = "" if unit.scale == 1 else cls.format_exponential_notation(unit.scale)
+
+        # dimensionless does not have any bases, but can have a scale;
+        # e.g., u.percent.decompose() gives "0.01".
+        if not unit.bases:
+            return s
+
+        if s:
+            s += cls._scale_unit_separator
+        # Unit powers are monotonically decreasing
+        if not fraction or unit.powers[-1] > 0:
+            return s + cls._format_unit_list(zip(unit.bases, unit.powers))
+
+        positive = []
+        negative = []
+        for base, power in zip(unit.bases, unit.powers, strict=True):
+            if power > 0:
+                positive.append((base, power))
             else:
-                s += cls._format_unit_list(numerator)
-
-        return s
+                negative.append((base, -power))
+        try:
+            return cls._fraction_formatters[fraction](
+                s,
+                cls._format_unit_list(positive) or "1",
+                cls._format_unit_list(negative),
+            )
+        except KeyError:
+            # We accept Booleans, but don't advertise them in the error message
+            *all_but_last, last = (
+                key for key in cls._fraction_formatters if isinstance(key, str)
+            )
+            supported_formats = (
```

Comment:
(For the `dict` option) I'd do
```
supported_formats = [repr(key) for key in cls._fraction_formatters if isinstance(key, str)]
supported_formats = ', '.join(supported[:-2] + [" or ".join(supported[-2:])])
```
(No oxford comma here?)

Moving the `repr` to the iteration may be good regardless.

p.s. This really seems something for the standard library, given the number of stackoverflow items. My solution turns out to be the same as https://stackoverflow.com/a/19839338, while a version with the oxford comma is at https://stackoverflow.com/a/58692463 (and yours just above the latter).

Reply:
Moving the `repr()` call was a good idea and shortened the code by a couple of lines.